core/server/worker/src/main/java/alluxio/worker/block/MonoBlockStore.java

```diff
@@ -301,44 +311,67 @@ public List<BlockStatus> load(List<Block> blocks, String tag, OptionalLong bandw
-      CompletableFuture<BlockStatus> future = RetryUtils.retryCallable("read from ufs",
+      CompletableFuture<Void> future = RetryUtils.retryCallable("read from ufs",
               () -> manager.read(blockId, block.getOffsetInFile(), blockSize, block.getUfsPath(),
                   false, tag), new ExponentialBackoffRetry(1000, 5000, 5))
+          // use orTimeout in java 11
+          .applyToEither(timeoutAfter(LOAD_TIMEOUT, TimeUnit.MILLISECONDS), d -> d)
           .thenAcceptAsync(d -> blockWriter.append(ByteBuffer.wrap(d)),
               GrpcExecutors.BLOCK_WRITER_EXECUTOR)
           .thenRun(() -> {
             try {
               blockWriter.close();
             } catch (IOException e) {
-              abortBlock(sessionId, block.getBlockId());
               throw AlluxioRuntimeException.from(e);
             }
           })
-          .thenRun(() -> commitBlock(sessionId, blockId, false)).handle((res, throwable) -> {
-            if (throwable == null) {
-              return null;
-            } else {
+          .thenRun(() -> commitBlock(sessionId, blockId, false))
+          .exceptionally(t -> {
+            AlluxioRuntimeException exception = AlluxioRuntimeException.from(t.getCause());
+            BlockStatus.Builder builder = BlockStatus.newBuilder().setBlock(block)
+                .setCode(exception.getStatus().getCode().value())
+                .setRetryable(exception.isRetryable());
+            if (exception.getMessage() != null) {
+              builder.setMessage(exception.getMessage());
+            }
+            errors.add(builder.build());
+            if (hasTempBlockMeta(blockId)) {
               try {
-                abortBlock(sessionId, block.getBlockId());
-              } catch (Exception e) {
-                throwable.addSuppressed(e);
+                abortBlock(sessionId, blockId);
+              } catch (Exception ee) {
+                LOG.warn(format("fail to abort temp block %s after failing to load block",
+                    block.getBlockId()), ee);
               }
-              AlluxioRuntimeException exception = AlluxioRuntimeException.from(throwable);
-              return BlockStatus.newBuilder().setBlock(block)
-                  .setCode(exception.getStatus().getCode().value())
-                  .setMessage(throwable.getMessage()).setRetryable(exception.isRetryable()).build();
             }
+            return null;
           });
       futures.add(future);
     }
-    errors.addAll(futures.stream().map(CompletableFuture::join).filter(Objects::nonNull)
-        .collect(Collectors.toList()));
-    return errors.build();
+    return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
+        .thenApply(x -> errors);
+  }
+
+  private <T> CompletableFuture<T> timeoutAfter(long timeout, TimeUnit unit) {
+    CompletableFuture<T> result = new CompletableFuture<>();
+    mDelayer.schedule(() -> result.completeExceptionally(new DeadlineExceededRuntimeException(
```

Comment:
Do we really have to have a separate thread just to do this 😂 I can't believe there's no better way 😂

Reply:
this is basically how java9 implements this feature